op-challenger/game/fault/trace/cannon/provider.go

```diff
@@ -142,6 +146,18 @@ func (p *CannonTraceProvider) loadProof(ctx context.Context, i uint64) (*proofDa
 		// If the requested index is after the last step in the actual trace, extend the final no-op step
 		return p.lastProof, nil
 	}
+	// Attempt to read the last step from disk cache
+	if p.lastProof == nil && p.lastStep == 0 {
+		step, err := ReadLastStep(p.dir)
+		if err != nil {
+			p.logger.Warn("Failed to read last step from disk cache", "err", err)
+		} else {
+			p.lastStep = step
+			if i > p.lastStep {
+				i = step
```

Comment:
Why do we lower `i` here?

Reply:
To avoid having to generate the proof again since the last step's proof should be present in this case.

Reply:
Specifically, the post state of the final step in the actual execution is extended to all remaining trace indices in the game.  So if we request a step after the end of the trace, we actually need to generate the proof at the final step cannon will execute and use that instead.